rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+
+We also want to make the API flexible enough to support additional algorithms
+that might be faster on certain hardware
+([6xBF16](https://arxiv.org/pdf/1904.06376.pdf), 3xTF32, etc) that can't fit
+into the limited precision_config values of `DEFAULT`, `HIGH` and `HIGHEST`. For
+example "6xBF16" provides similar precision to F32, but about 2x performance on
+some GPUs.
+
+When an algorithm is not supported by an accelerator, the program should fail
+instead of implicitly falling back to another. It should be the responsibility
+of higher level frameworks such as JAX, to define the required algorithm for
+each accelerator that is used.
+
+## Proposed Specification change
+
+### dot_general
+
+#### Semantics
+
+*Add these lines:*
+
+`algorithm` defines the main properties of the algorithm used to implement the
+dot operation, which also defines the precision. If the algorithm value is set
+(`!= UNSET`), then the `precision_config` is ignored. If the algorithm value is
+not set (`= UNSET`), then the `precision_config` will take effect.
+
+In general the first two types in the algorithm name are `AType` and `BType` -
+the precisions that the LHS and RHS of the operation are rounded to, and the
+third one is `AccumType` - the accumulation type. These types are independent
+from the storage types of the inputs and the output (but not all combinations
+are supported). Some algorithms have other suffixes too.
+
+This is the current list of algorithms:
+
+* `UNSET`: There is no restriction on the algorithm, the precision is decided
+  based on the `precision_config`. Use this for every type combination that
+  doesn't have a specific algorithm.
+* `ANY_F8_ANY_F8_F32`: The inputs are used with the full precision of their
+  storage type (which must be an 8-bit floating point type - the same for both
+  operands). The accumulator type is F32.
+* `ANY_F8_ANY_F8_F32_FAST_ACCUM`: The inputs are used with the full precision of
+    their storage type (which must be an 8-bit floating point type - the same
+    for both operands). The accumulator type is formally F32, but the
+    intermediate results will not periodically be promoted to a higher
+    precision.
+* `F16_F16_F16`: The inputs are used with / casted to F16 precision and the
+  accumulator type is also F16.
+* `F16_F16_F32`: The inputs are used with / casted to F16 precision and the
+  accumulator type is F32.
+* `BF16_BF16_BF16`: The inputs are used with / casted to BF16 precision and the
+  accumulator type is also BF16.
+* `BF16_BF16_F32`: The inputs are used with / casted to BF16 precision and the
+  accumulator type is F32.
+* `BF16_BF16_F32_X3`: An algorithm which uses three BF16_BF16_F32 dot operations
+  per "tile" to achive a higher precision.
+* `BF16_BF16_F32_X6`: An algorithm which uses six BF16_BF16_F32 dot operations
+  per "tile" to achive a higher precision ("similar" to F32_F32_F32).
+* `TF32_TF32_F32`: The inputs are casted to TF32 precision and the accumulator
```

Comment:
If this text is proposed, IMO, we must have a separate section to describe what `TF32` means (and the differences between `F32` and `TF32`).

Reply:
It was defined by NVIDIA, is it enough to link to this?
https://blogs.nvidia.com/blog/tensorfloat-32-precision-format/